llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2221,72 +2221,79 @@ UnPackOp::getTiledImplementation(OpBuilder &builder,
       outputExpandedSizes;
   for (auto dim : llvm::seq<int64_t>(0, outputRank)) {
-    if (dimAndTileMapping.count(dim)) {
-      DivModValue firstCoord =
-          getDivMod(builder, loc,
-                    getValueOrCreateConstantIndexOp(builder, loc, offsets[dim]),
-                    getValueOrCreateConstantIndexOp(builder, loc,
-                                                    dimAndTileMapping[dim]));
-      DivModValue lastCoord = getDivMod(
-          builder, loc,
-          getValueOrCreateConstantIndexOp(
-              builder, loc, sub(add(offsets[dim], sizes[dim]), oneAttr)),
-          getValueOrCreateConstantIndexOp(builder, loc,
-                                          dimAndTileMapping[dim]));
-
-      inputIndices.push_back(firstCoord.quotient);
-
-      // Get the upper bound because it could be an extract_slice case. The
-      // sizes are determined by loop bound and step, where loop bound is the
-      // size of output shape.
-      // In incomplete tile cases, the input could have larger shape, it is safe
-      // to extend the boundary because they are pre-padded. I.e., the size of
-      // input dim is always aligned to inner_tile_size.
-      FailureOr<int64_t> cstSize = linalg::getConstantUpperBoundForIndex(
-          getValueOrCreateConstantIndexOp(builder, loc, sizes[dim]));
-      Optional<int64_t> cstInnerSize =
-          getConstantIntValue(dimAndTileMapping[dim]);
-      if (!failed(cstSize) && cstInnerSize &&
-          cstSize.value() % cstInnerSize.value() == 0) {
-        inputSizes.push_back(
-            builder.getIndexAttr(cstSize.value() / cstInnerSize.value()));
-      } else {
-        inputSizes.push_back(
-            add(sub(lastCoord.quotient, firstCoord.quotient), oneAttr));
-      }
-      outputNewOffsets.push_back(firstCoord.remainder);
-
-      AffineExpr i, tile;
-      bindDims(builder.getContext(), i);
-      bindSymbols(builder.getContext(), tile);
-      OpFoldResult size = makeComposedFoldedAffineApply(
-          builder, loc, i * tile,
-          ArrayRef<OpFoldResult>{inputSizes.back(), dimAndTileMapping[dim]});
-      outputExpandedSizes.push_back(size);
-    } else {
+    if (!dimAndTileMapping.count(dim)) {
       inputIndices.push_back(offsets[dim]);
       inputSizes.push_back(sizes[dim]);
       outputNewOffsets.push_back(zeroAttr);
       outputExpandedSizes.push_back(sizes[dim]);
+      continue;
     }
+
+    // Get the upper bound because the "output_size * inner_tile_size" could
+    // be less than input size. The sizes are determined by loop bound and
+    // step, where loop bound is the size of output shape. In incomplete tile
+    // cases, the input could have larger shape, it is safe to extend the
+    // boundary because they are pre-padded. I.e., the size of input dim is
+    // always aligned to inner_tile_size.
+    FailureOr<int64_t> cstSize = linalg::getConstantUpperBoundForIndex(
```

Comment:
We don't care what the static sizes are here and only that they're divisible. If we're not tracking somewhere places that are assuming static values we should do that (could just be a `TODO(#issue)` or something) so in the short term this kind of logic can be factored out into an `isAlignedTo(Value, Value)` that we can eventually plug in the analysis without needing to scour the codebase.

Reply:
Yes, it is an optimization or say a short cut.

After thinking a while, I'm not going to factor them out into an `isAlignedTo` method because it's assuming the size/offset are from tiling artifacts. Ideally, these affine_ops should be folded into a constant; they do for most of cases. unpack op is quite special because the affine ops are complicated. Thus we added an unpack op specific optimization here.